aws-lc-rs/src/pqdsa/key_pair.rs

```diff
@@ -0,0 +1,284 @@
+    }
+}
+
+/// A PQDSA private key.
+pub struct PqdsaPrivateKey<'a>(pub(crate) &'a PqdsaKeyPair);
+
+impl AsDer<Pkcs8V1Der<'static>> for PqdsaPrivateKey<'_> {
+    /// Serializes the key to PKCS#8 v1 DER.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` if serialization fails.
+    fn as_der(&self) -> Result<Pkcs8V1Der<'static>, Unspecified> {
+        Ok(Pkcs8V1Der::new(
+            self.0
+                .evp_pkey
+                .marshal_rfc5208_private_key(pkcs8::Version::V1)?,
+        ))
+    }
+}
+
+impl AsRawBytes<PqdsaPrivateKeyRaw<'static>> for PqdsaPrivateKey<'_> {
+    fn as_raw_bytes(&self) -> Result<PqdsaPrivateKeyRaw<'static>, Unspecified> {
+        Ok(PqdsaPrivateKeyRaw::new(
+            self.0.evp_pkey.marshal_raw_private_key()?,
+        ))
+    }
+}
+
+impl PqdsaKeyPair {
+    /// Generates a new PQDSA key pair for the specified algorithm.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` is the key generation fails.
+    pub fn generate(algorithm: &'static PqdsaSigningAlgorithm) -> Result<Self, Unspecified> {
+        let evp_pkey = evp_key_pqdsa_generate(algorithm.0.id.nid())?;
+        let pubkey = PublicKey::from_private_evp_pkey(&evp_pkey)?;
+        Ok(Self {
+            algorithm,
+            evp_pkey,
+            pubkey,
+        })
+    }
+
+    /// Parses a PKCS#8 key from the specified bytes.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` if the key is not valid for the specified signing algorithm.
+    pub fn from_pkcs8(
+        algorithm: &'static PqdsaSigningAlgorithm,
+        pkcs8: &[u8],
+    ) -> Result<Self, KeyRejected> {
+        let evp_pkey = LcPtr::<EVP_PKEY>::parse_rfc5208_private_key(pkcs8, EVP_PKEY_PQDSA)?;
+        validate_pqdsa_evp_key(&evp_pkey, algorithm.0.id)?;
+        let pubkey = PublicKey::from_private_evp_pkey(&evp_pkey)?;
+        Ok(Self {
+            algorithm,
+            evp_pkey,
+            pubkey,
+        })
+    }
+
+    /// Constructs a key pair from a raw private key.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` if the key is not valid for the specified signing algorithm.
+    pub fn from_raw_private_key(
+        algorithm: &'static PqdsaSigningAlgorithm,
+        raw_private_key: &[u8],
+    ) -> Result<Self, KeyRejected> {
+        let evp_pkey = LcPtr::<EVP_PKEY>::parse_raw_private_key(raw_private_key, EVP_PKEY_PQDSA)?;
+        validate_pqdsa_evp_key(&evp_pkey, algorithm.0.id)?;
+        let pubkey = PublicKey::from_private_evp_pkey(&evp_pkey)?;
+        Ok(Self {
+            algorithm,
+            evp_pkey,
+            pubkey,
+        })
+    }
+
+    /// Serializes the private key to PKCS#8 v1 DER.
+    ///
+    /// # Errors
+    /// Returns `Unspecified` if serialization fails.
+    pub fn to_pkcs8(&self) -> Result<Document, Unspecified> {
+        Ok(Document::new(
+            self.evp_pkey.marshal_rfc5208_private_key(Version::V1)?,
+        ))
+    }
+
+    /// Uses this key to sign the message provided.
```

Comment:
Nit: Could be more clearer on what's being returned and where the signature is written to on success